src/environment.py

```diff
@@ -0,0 +1,298 @@
+
+    if config.get("s3_server_side_encryption") and not config.get("license"):
+        missing.append("license")
+
+    if config.get("sso") == "true":
+        missing += [setting for setting in REQUIRED_SSO_SETTINGS if not config.get(setting)]
+
+    return sorted(missing)
+
+
+def generate(config: dict, app_name: str, site_url: str, db_uri: str) -> dict:
+    """Generate container app environment based on charm configuration.
+
+    Args:
+        config: dict of the charm's configuration
+        app_name: name of the app
+        site_url: public facing URL of mattermost
+        db_uri: URI of the psql database
+
+    Returns:
+        dict of container app environment.
+    """
+    # https://github.com/mattermost/mattermost-server/pull/14666
+    db_uri = db_uri.replace("postgresql://", "postgres://")
+    env = {
+        "MATTERMOST_HTTPD_LISTEN_PORT": str(CONTAINER_PORT),
+        "MM_CONFIG": db_uri,
+        "MM_SQLSETTINGS_DATASOURCE": db_uri,
+        # image proxy
+        "MM_IMAGEPROXYSETTINGS_ENABLE": "true" if config["image_proxy_enabled"] else "false",
+        "MM_IMAGEPROXYSETTINGS_IMAGEPROXYTYPE": "local",
+        # logging
+        "MM_LOGSETTINGS_CONSOLELEVEL": "DEBUG" if config["debug"] else "INFO",
+        "MM_LOGSETTINGS_ENABLECONSOLE": "true",
+        "MM_LOGSETTINGS_ENABLEFILE": "false",
+        "MM_TEAMSETTINGS_MAXCHANNELSPERTEAM": config["max_channels_per_team"],
+        "MM_TEAMSETTINGS_MAXUSERSPERTEAM": config["max_users_per_team"],
+        "MM_SERVICESETTINGS_SITEURL": site_url,
+    }
+
+    if config["primary_team"]:
+        env["MM_TEAMSETTINGS_EXPERIMENTALPRIMARYTEAM"] = config["primary_team"]
+
+    if config["outbound_proxy"]:
+        env["HTTP_PROXY"] = config["outbound_proxy"]
+        env["HTTPS_PROXY"] = config["outbound_proxy"]
+        if config["outbound_proxy_exceptions"]:
+            env["NO_PROXY"] = config["outbound_proxy_exceptions"]
+
+    if config["s3_enabled"]:
+        env.update(
+            {
+                "MM_FILESETTINGS_DRIVERNAME": "amazons3",
+                "MM_FILESETTINGS_MAXFILESIZE": config["max_file_size"] * 1048576,  # LP:1881227
+                "MM_FILESETTINGS_AMAZONS3SSL": "true",  # defaults to true; belt and braces
+                "MM_FILESETTINGS_AMAZONS3ENDPOINT": config["s3_endpoint"],
+                "MM_FILESETTINGS_AMAZONS3BUCKET": config["s3_bucket"],
+                "MM_FILESETTINGS_AMAZONS3REGION": config["s3_region"],
+                "MM_FILESETTINGS_AMAZONS3ACCESSKEYID": config["s3_access_key_id"],
+                "MM_FILESETTINGS_AMAZONS3SECRETACCESSKEY": config["s3_secret_access_key"],
+                "MM_FILESETTINGS_AMAZONS3SSE": "true"
+                if config["s3_server_side_encryption"]
+                else "false",
+                "MM_FILESETTINGS_AMAZONS3TRACE": "true" if config["debug"] else "false",
+            }
+        )
+
+    env.update(_env_for_canonical_defaults(config))
```

Comment:
Something to consider, all of these could just yield tuples which are passed to `update` as required, would save generating and merging a bunch of dictinaries

Reply:
Changed to generator